File: app/src/main/java/com/example/grocer21/MainActivity.java
package com.example.grocer21;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.NavigationView;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;

import com.example.grocer21.Database.Allergy;
import com.example.grocer21.Database.Diet;
import com.example.grocer21.Database.Food;

import java.util.List;
import java.util.Objects;

public class MainActivity extends AppCompatActivity
        implements
        FoodFragment.OnListFragmentInteractionListener,
        HomeScreenFragment.OnFragmentInteractionListener,
        AllergyFragment.OnListFragmentInteractionListener,
        DietFragment.OnListFragmentInteractionListener,
        FoodItemFragment.OnFragmentInteractionListener,
        DietItemFragment.OnFragmentInteractionListener,
        AllergyItemFragment.OnFragmentInteractionListener,
        AllergyListContainer.OnFragmentInteractionListener,
        FoodListContainer.OnFragmentInteractionListener,
        DietListContainer.OnFragmentInteractionListener {

    private DrawerLayout mDrawerLayout;
    private Fragment fragment = null;
    private final FragmentManager fragmentManager = getSupportFragmentManager();
    private boolean onTopLevelNav = true;
    private Toolbar toolbar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        ActionBar actionbar = getSupportActionBar();
        Objects.requireNonNull(actionbar).setDisplayHomeAsUpEnabled(true);
        actionbar.setHomeAsUpIndicator(R.drawable.ic_menu);

        mDrawerLayout = findViewById(R.id.drawer_layout);

        fragment = new HomeScreenFragment();
        fragmentManager.beginTransaction().add(R.id.fragmentContainer, fragment).commit();

        final DatabaseViewModel databaseViewModel = ViewModelProviders.of(this).get(DatabaseViewModel.class);
        databaseViewModel.allFoods.observe(this, new Observer<List<Food>>() {
            @Override
            public void onChanged(@Nullable List<Food> foods) {

            }
        });
        databaseViewModel.allAllergies.observe(this, new Observer<List<Allergy>>() {
            @Override
            public void onChanged(@Nullable List<Allergy> allergies) {

            }
        });
        databaseViewModel.allDiets.observe(this, new Observer<List<Diet>>() {
            @Override
            public void onChanged(@Nullable List<Diet> diets) {

            }
        });

        NavigationView navigationView = findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(
                new NavigationView.OnNavigationItemSelectedListener() {
                    @Override
                    public boolean onNavigationItemSelected(@NonNull MenuItem menuItem) {
                        int id = menuItem.getItemId();

                        if (id == R.id.nav_food_list) {
                            fragment = new FoodListContainer();
                            toolbar.setTitle("Food");
                        } else if (id == R.id.nav_allergy_list) {
                            fragment = new AllergyListContainer();
                            toolbar.setTitle("Allergies");
                        } else if (id == R.id.nav_diet_list) {
                            fragment = new DietListContainer();
                            toolbar.setTitle("Diet");
                        }

                        setTitle(menuItem.getTitle());
                        mDrawerLayout.closeDrawers();
                        if (onTopLevelNav) {
                            if (fragment != null) {
                                fragmentManager.beginTransaction().replace(R.id.fragmentContainer, fragment).addToBackStack(null).commit();
                                onTopLevelNav = false;
                            }
                        } else {
                            fragmentManager.beginTransaction().replace(R.id.fragmentContainer, fragment).commit();
                        }

                        return true;
                    }
                });

        mDrawerLayout.addDrawerListener(
                new DrawerLayout.DrawerListener() {
                    @Override
                    public void onDrawerSlide(@NonNull View drawerView, float slideOffset) {
                        // Respond when the drawer's position changes
                    }

                    @Override
                    public void onDrawerOpened(@NonNull View drawerView) {
                        // Respond when the drawer is opened
                    }

                    @Override
                    public void onDrawerClosed(@NonNull View drawerView) {
                        // Respond when the drawer is closed
                    }

                    @Override
                    public void onDrawerStateChanged(int newState) {
                        // Respond when the drawer motion state changes
                    }
                }
        );
    }

    public MainActivity() {
        toolbar = null;
    }


    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                mDrawerLayout.openDrawer(GravityCompat.START);
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onListFragmentInteraction(Food item) {

        FragmentManager fragmentManager = getSupportFragmentManager();
        String name = item.getName();
        toolbar.setTitle(name);
        Bundle bundle = new Bundle();
        bundle.putString("name", name);
        FoodItemFragment foodItemFragment = new FoodItemFragment();
        foodItemFragment.setArguments(bundle);
        fragmentManager.beginTransaction()
                .replace(R.id.fragmentContainer, foodItemFragment)
                .addToBackStack(null)
                .commit();
        onTopLevelNav = false;

    }

    public void displayHelp(View view) {
        DialogFragment newHelpFragment = new HelpDialogFragment();
        newHelpFragment.show(getSupportFragmentManager(), "HelpDialog");
    }

    @Override
    public void onListFragmentInteraction(Allergy item) {

        FragmentManager fragmentManager = getSupportFragmentManager();
        String name = item.getName();
        toolbar.setTitle(name);
        Bundle bundle = new Bundle();
        bundle.putString("name", name);
        AllergyItemFragment allergyItemFragment = new AllergyItemFragment();
        allergyItemFragment.setArguments(bundle);
        fragmentManager.beginTransaction()
                .replace(R.id.fragmentContainer, allergyItemFragment)
                .addToBackStack(null)
                .commit();
        onTopLevelNav = false;

    }

    @Override
    public void onListFragmentInteraction(Diet item) {

        FragmentManager fragmentManager = getSupportFragmentManager();
        String name = item.getName();
        toolbar.setTitle(name);
        Bundle bundle = new Bundle();
        bundle.putString("name", name);
        DietItemFragment dietItemFragment = new DietItemFragment();
        dietItemFragment.setArguments(bundle);
        fragmentManager.beginTransaction()
                .replace(R.id.fragmentContainer, dietItemFragment)
                .addToBackStack(null)
                .commit();
        onTopLevelNav = false;

    }

    public void onBackPressed() {
        FragmentManager fragmentManager = getSupportFragmentManager();
        if (fragmentManager.getBackStackEntryCount() >= 0) {
            fragment = new HomeScreenFragment();
            fragmentManager.beginTransaction()
                    .replace(R.id.fragmentContainer, fragment)
                    .addToBackStack(null)
                    .commit();
            onTopLevelNav = true;
        }else {
            super.onBackPressed();
        }
    }

    @Override
    public void onFragmentInteraction() {

    }

    @Override
    public void onFragmentInteraction(Uri uri) {

    }
}

File: app/src/main/java/com/example/grocer21/MainActivity.kt
package com.example.grocer21

import android.net.Uri
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModel
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.grocer21.database.entities.Allergies
import com.example.grocer21.database.entities.Diets
import com.example.grocer21.database.entities.Foods
import com.google.android.material.navigation.NavigationView

/**
 * The main activity that houses the fragments and handles interactions between them
 */
class MainActivity : AppCompatActivity(),
        FoodListFragment.OnFoodListFragmentInteractionListener,
        HomeScreenFragment.OnHomeScreenFragmentInteractionListener,
        AllergyListFragment.OnAllergyListFragmentInteractionListener,
        DietListFragment.OnDietListFragmentInteractionListener,
        FoodItemFragment.OnFoodItemFragmentInteractionListener,
        DietItemFragment.OnDietItemFragmentInteractionListener,
        AllergyItemFragment.AllergyItemFragmentInteractionListener,
        AllergyListContainer.OnAllergyListContainerInteractionListener,
        FoodListContainer.OnFoodListContainerClickListener,
        DietListContainer.OnDietListContainerInteractionListener,
        AddNewFood.OnSaveButtonPressedListener {
    override fun onDietListContainerInteraction(uri: Uri) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun onDietItemFragmentInteraction() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun onFoodItemFragmentInteraction() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun onHomeScreenFragmentInteraction() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }


    private var mDrawerLayout: DrawerLayout? = null
    private var fragment: Fragment? = null
    private val fragmentManager = supportFragmentManager
    private var onTopLevelNav = true
    private var toolbar: Toolbar? = null
    private lateinit var databaseViewModel: ViewModel


    /**
     * Lifecycle method
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)


        toolbar = findViewById(R.id.toolbar)
        setSupportActionBar(toolbar)
        val actionbar = supportActionBar
        (actionbar)?.setDisplayHomeAsUpEnabled(true)
        (actionbar ?: return).setHomeAsUpIndicator(R.drawable.ic_menu)

        mDrawerLayout = findViewById(R.id.drawer_layout)

        fragment = HomeScreenFragment()
        fragmentManager.beginTransaction().add(R.id.fragmentContainer, fragment ?: return).commit()

        databaseViewModel = DatabaseViewModel(application)

        if (findViewById<RecyclerView>(R.id.foodRecyclerView) != null) {
            val recyclerView = findViewById<RecyclerView>(R.id.foodRecyclerView) ?: return
            val foodAdapter = FoodListAdapter(this)
            recyclerView.adapter = foodAdapter
            recyclerView.layoutManager = LinearLayoutManager(this)

            (databaseViewModel as DatabaseViewModel).allFoods.observe(this, Observer { foods ->
                foods?.let { foodAdapter.setFoods(it) }
            })
        }

        if (findViewById<RecyclerView>(R.id.allergyRecyclerView) != null) {
            val recyclerView = findViewById<RecyclerView>(R.id.allergyRecyclerView) ?: return
            val allergyAdapter = AllergyListAdapter(this)
            recyclerView.adapter = allergyAdapter
            recyclerView.layoutManager = LinearLayoutManager(this)

            (databaseViewModel as DatabaseViewModel)
                    .allAllergies
                    .observe(this,
                            Observer { allergies ->
                                allergies?.let { allergyAdapter.setAllergies(it) }
                            })
        }

        if (findViewById<RecyclerView>(R.id.dietRecyclerView) != null) {
            val recyclerView = findViewById<RecyclerView>(R.id.dietRecyclerView) ?: return
            val dietAdapter = DietListAdapter(this)
            recyclerView.adapter = dietAdapter
            recyclerView.layoutManager = LinearLayoutManager(this)

            (databaseViewModel as DatabaseViewModel).allDiets.observe(this, Observer { diets ->
                diets?.let { dietAdapter.setDiets(it) }
            })
        }

        val navigationView = findViewById<NavigationView>(R.id.nav_view)
        navigationView.setNavigationItemSelectedListener { menuItem ->

            when (menuItem.itemId) {
                R.id.nav_food_list -> {
                    fragment = FoodListContainer()
                    toolbar!!.title = "Foods"
                }
                R.id.nav_allergy_list -> {
                    fragment = AllergyListContainer()
                    toolbar!!.title = "Allergies"
                }
                R.id.nav_diet_list -> {
                    fragment = DietListContainer()
                    toolbar!!.title = "Diet"
                }
            }

            title = menuItem.title
            mDrawerLayout!!.closeDrawers()
            if (onTopLevelNav) {
                if (fragment != null) {
                    fragmentManager
                            .beginTransaction()
                            .replace(R.id.fragmentContainer, fragment!!)
                            .addToBackStack(null)
                            .commit()
                    onTopLevelNav = false


                }
            } else {
                fragmentManager
                        .beginTransaction()
                        .replace(R.id.fragmentContainer, fragment!!)
                        .commit()
            }

            true
        }

        (mDrawerLayout ?: return).addDrawerListener(
                object : DrawerLayout.DrawerListener {
                    override fun onDrawerSlide(drawerView: View, slideOffset: Float) {
                        // Respond when the drawer's position changes
                    }

                    override fun onDrawerOpened(drawerView: View) {
                        // Respond when the drawer is opened
                    }

                    override fun onDrawerClosed(drawerView: View) {
                        // Respond when the drawer is closed
                    }

                    override fun onDrawerStateChanged(newState: Int) {
                        // Respond when the drawer motion state changes
                    }
                }
        )
    }

    init {
        toolbar = null
    }


    /**
     * function that closes the navigation drawer if an item was selected
     */
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                mDrawerLayout!!.openDrawer(GravityCompat.START)
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onFoodListFragmentInteraction(item: Foods) {

        val fragmentManager = supportFragmentManager
        val name = item.getName()
        (toolbar ?: return).title = name
        val bundle = Bundle()
        bundle.putString("name", name)
        val foodItemFragment = FoodItemFragment()
        foodItemFragment.arguments = bundle
        fragmentManager.beginTransaction()
                .replace(R.id.fragmentContainer, foodItemFragment)
                .addToBackStack(null)
                .commit()
        onTopLevelNav = false

    }

    /**
     * creates a dialog fragment when the "help" button is pressed (see [HelpDialogFragment])
     *
     * view variable is there for the onClick attach in the XML file
     */
    fun displayHelp(view: View) {
        val newHelpFragment = HelpDialogFragment()
        newHelpFragment.show(supportFragmentManager, "HelpDialog")
    }

    override fun onListFragmentInteraction(item: Allergies) {

        val fragmentManager = supportFragmentManager
        val name = item.getName()
        (toolbar ?: return).title = name
        val bundle = Bundle()
        bundle.putString("name", name)
        val allergyItemFragment = AllergyItemFragment()
        allergyItemFragment.arguments = bundle
        fragmentManager.beginTransaction()
                .replace(R.id.fragmentContainer, allergyItemFragment)
                .addToBackStack(null)
                .commit()
        onTopLevelNav = false

    }

    override fun onDietListFragmentInteraction(item: Diets) {

        val fragmentManager = supportFragmentManager
        val name = item.getName()
        (toolbar ?: return).title = name
        val bundle = Bundle()
        bundle.putString("name", name)
        val dietItemFragment = DietItemFragment()
        dietItemFragment.arguments = bundle
        fragmentManager.beginTransaction()
                .replace(R.id.fragmentContainer, dietItemFragment)
                .addToBackStack(null)
                .commit()
        onTopLevelNav = false

    }

    /**
     * Navigation method that handles back button presses
     *
     * will be modified when actual navigation structure is implemented
     */
    override fun onBackPressed() {
        val fragmentManager = supportFragmentManager
        if (fragmentManager.backStackEntryCount >= 0) {
            fragment = HomeScreenFragment()
            fragmentManager.beginTransaction()
                    .replace(R.id.fragmentContainer, fragment ?: return)
                    .addToBackStack(null)
                    .commit()
            onTopLevelNav = true
        } else {
            super.onBackPressed()
        }
    }

    override fun onAllergyFragmentInteraction() {

    }

    override fun onAllergyListContainerInteraction(uri: Uri) {

    }

    override fun onClick(v: View) {
        fragment = AddNewFood()
        supportFragmentManager.beginTransaction()
                .replace(R.id.fragmentContainer, fragment ?: return)
                .addToBackStack(null)
                .commit()
    }

    override fun onNewFoodSaved(position: Bundle) {
        val newFood = Foods(position.getLong("upc"), position.getString("name"))
        (databaseViewModel as DatabaseViewModel).insertWrapper(newFood)
    }

    /**
     * function that overrides [AddNewFood.setOnSaveButtonPressedListener]
     */
    override fun onAttachFragment(fragment: Fragment) {
        if (fragment is AddNewFood) {
            fragment.setOnSaveButtonPressedListener(this)
        }
    }
}

